src/main/java/org/broadinstitute/hellbender/dev/pipelines/bqsr/BQSR_Dataflow.java

```diff
@@ -0,0 +1,276 @@
+    return new String[] { fastaFilename, ReferenceUtils.getFastaDictionaryFileName(fastaFilename), ReferenceUtils.getFastaIndexFileName(fastaFilename) };
+  }
+
+  // a key to group intervals by (aka sharding)
+  private static String posKey(Locatable loc) {
+    return loc.getContig() + (loc.getStart() / BLOCKSIZE);
+  }
+
+  // a key to group intervals by (aka sharding)
+  private static String posKey(Read r) {
+    // Reads are 0-based, SimpleIntervals are 1-based
+    int start = r.getAlignment().getPosition().getPosition().intValue() + 1;
+    return posKey(new SimpleInterval(r.getAlignment().getPosition().getReferenceName(),
+            start,
+            start));
+  }
+
+  // the shard that the last base of the interval lands in,
+  // if different from the first base's. Null otherwise.
+  private static String posKeyEnd(Locatable loc) {
+    if (loc.getStart()/BLOCKSIZE == (loc.getEnd())/BLOCKSIZE) return null;
+    return loc.getContig() + (loc.getEnd() / BLOCKSIZE);
+  }
+
+  /**
+   * returns union(reads,placestoIgnore).groupBy(x->posKey(x))
+   */
+  private static PCollection<KV<String, CoGbkResult>> groupByBlock(PCollection<Read> reads, PCollection<SimpleInterval> placesToIgnore) {
+    // shard reads
+    PCollection<KV<String, Read>> shardedReads = reads.apply(ParDo
+        .named("shard reads")
+        .of(
+            new DoFn<Read, KV<String, Read>>() {
+              @Override
+              public void processElement(ProcessContext c) {
+                Read r = c.element();
+                c.output(KV.of(posKey(r), r));
+              }
+
+            }))
+        // Dataflow boilerplate
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), GenericJsonCoder.of(Read.class)));
+    // send ignores to every shard that overlaps with them (so, possibly more than one).
+    // (for now we assume they can't span more than two)
+    PCollection<KV<String, SimpleInterval>> shardedIgnore = placesToIgnore.apply(ParDo
+        .named("shard known intervals")
+        .of(
+            new DoFn<SimpleInterval, KV<String, SimpleInterval>>() {
+              @Override
+              public void processElement(ProcessContext c) {
+                SimpleInterval i = c.element();
+                String firstPos = posKey(i);
+                c.output(KV.of(firstPos, i));
+                String secondPos = posKeyEnd(i);
+                if (null!=secondPos) {
+                  c.output(KV.of(secondPos, i));
+                }
+              }
+            }))
+        // Dataflow boilerplate
+        .setCoder(KvCoder.of(StringUtf8Coder.of(), SerializableCoder.of(SimpleInterval.class)));
+
+    return KeyedPCollectionTuple.of(readTag, shardedReads)
+        .and(intervalTag, shardedIgnore)
+        .apply(CoGroupByKey.<String>create());
+  }
+
+  /**
+   * For each input block, compute its statistics.
+   * <p>
+   * Basically just delegate to the CalibrationTablesBuilder.
```

Comment:
I can add that it also downloads the reference and prints statistics to the DF logs.